contents/javascript/Javascript-framework-Redux.md

```diff
@@ -0,0 +1,231 @@
+
+The reducers are <tooltip content="A pure function is a function which: Given the same input, will always return the same output. And produces no side effects.">**pure functions**</tooltip>. They take the previous state and an action, and return the next state as new state objects, instead of mutating the previous state.
+
+A careful eye would notice that the above code block examples on reducers are not pure functions. If you're interested, you may see how the reducer should actually be written in the [official Redux site](https://redux.js.org/basics/reducers/#source-code). The code in the link includes some Redux API, and Javascript methods such as [`.map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) and the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) to make the functions pure. But knowing the actual syntax for Redux is not the main focus of this introductory article.
+
+## Why Use Redux?
+
+<box type='definition' header="Some points given in this section were adapted from these articles">
+
+* [Why use Redux? Reasons with clear examples by Neo Ighodaro](https://blog.logrocket.com/why-use-redux-reasons-with-clear-examples-d21bffd5835/)
+* [Redux · An Introduction by Alex Bachuk](https://www.smashingmagazine.com/2016/06/an-introduction-to-redux/)
+</box>
+
+Now that we know what Redux is, let us take a look at how the structure of Redux and its concepts make possible some benefits.
+
+1. **State is made predictable**<br/>
+In Redux, the single immutable store is the single source of truth for application state. Neither the view of the application, nor network callbacks directly write to it. Changes to the state are via actions, which Redux centralizes and keep in strict order. Also, reducers are pure functions, always producing the same result for the same state and action. Because of all of this, there is a structured and predictable cycle of which applicable state is updated, making state is predictable. Having a predictable state, makes it possible to implement functionality that would be otherwise difficult, such as Undo and Redo.
+
+2. **Easier to debug and inspect**<br/>
+Since actions and state are plain objects, it is possible to serialize them and log them. In Redux, especially with the [Redux DevTools](https://github.com/reduxjs/redux-devtools), you can view a log of actions that took place, making it easy to understand errors in code. You can also print the store <tooltip content="for example using `console.log(store.getState())`">within the code itself</tooltip>. The Redux DevTools, used in the browser, also provides features such as a tree visualization of the application state, as well a time travel ability to replay state changes.
```

Comment:
`In Redux ...` -> `Redux DevTools provides first class support for application debugging, ...`